### PR TITLE
[train+tune] Make path joining OS-agnostic by using `Path.as_posix` over `os.path.join`

### PR DESCRIPTION
**Pull request by n30111**
## Why are these changes needed?

`os.path.join` uses an OS-dependent separator (`\` on windows and `/` on posix systems), which causes some issues when `os.path.join` is used in conjunction with `Path`. The combination can result in a mix of `/` and `\` when running Ray Train/Tune on windows. This runs into issues when passing these paths into `pyarrow.fs`, leading to issues such as `FileNotFound`.

This PR fixes the following issues:

While running training on Windows OS with S3 as storage system, some files are not placed in correct locations.
For example the file `checkpoint_00000/.metadata.json` ends up as `checkpoint_00000\.metadata.json`.
Also while restoring a tuner run, it fails, as `tuner.pkl` is not located correctly due to path issue. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/_checkpoint.py [hunk 1]
```diff
@@ -129,7 +130,7 @@ def get_metadata(self) -> Dict[str, Any]:
 
         If no metadata is stored, an empty dict is returned.
         """
-        metadata_path = os.path.join(self.path, _METADATA_FILE_NAME)
+        metadata_path = (Path(self.path) / _METADATA_FILE_NAME).as_posix()
```
Comment (justinvyu) [inline python/ray/train/_checkpoint.py:133]: Nit: I used to like the `/`, but I now think all the parentheses makes it annoying, so I prefer `Path(..., ...)` or `Path(...).joinpath(...).as_posix()` 😄 

```suggestion
        metadata_path = Path(self.path,  _METADATA_FILE_NAME).as_posix()
```

**Review by justinvyu — COMMENTED** [review]
Thanks for the contribution! I think we should generally use `Path.as_posix` everywhere we need to do path joins to be OS agnostic, and pyarrow (which we use under the hood) does not like the mix of posix and windows separators.

I'm wondering if we want to actually go ahead and change ALL of them with this PR? I did a global search on the repo and have found a bunch more instances of `os.path.join` in our Ray Train/Tune code.

Another note: we basically don't have comprehensive windows tests, so support for windows for at least Ray Train/Tune is *very* experimental. If you are not tied down to local development on a windows machine, running on a devbox or a linux subsystem may be a good thing to consider.

**Comment by n30111** [comment]
@justinvyu thanks, we are dependent on both linux and windows, so this changes will be useful.

**Comment by n30111** [comment]
@justinvyu I have updated the PR, It would be great if we can include these changes in the next release, thanks.

**Comment by justinvyu** [comment]
Hey @n30111 sorry for the delay, I'll do another review on this PR.

**Comment by jbedorf** [comment]
@justinvyu  Friendly ping on this PR.

**Comment by justinvyu** [comment]
@n3011 @jbedorf Is it possible to "give maintainers permission to edit" the PR (should be some setting in the PR)? I've replaced `os.path.join` in many other place sin the code.

**Comment by n30111** [comment]
@justinvyu  Hi Justin, that particular option does not work for forks hosted by organizations (see: https://github.com/orgs/community/discussions/5634 ), but we just added you to the fork/repository to enable you to push the changes you made.

**Comment by justinvyu** [comment]
@n3011 Is it possible to add a small test to confirm that this PR fixes the failure scenario you're running into? Thanks!

https://github.com/ray-project/ray/blob/master/python/ray/train/tests/test_windows.py

File: python/ray/tune/logger/json.py [hunk 2]
```diff
@@ -41,8 +41,8 @@ class JsonLogger(Logger):
 
     def _init(self):
         self.update_config(self.config)
-        local_file = os.path.join(self.logdir, EXPR_RESULT_FILE)
-        self.local_out = open(local_file, "a")
+        local_file = Path(self.logdir, EXPR_RESULT_FILE)
```
Comment (matthewdeng) [inline python/ray/tune/logger/json.py:44]: Should this include `as_posix()`?

Comment (matthewdeng) [reply python/ray/tune/logger/json.py:44]: Same question for other instances below.

**Review by matthewdeng — APPROVED** [review]

File: python/ray/train/_checkpoint.py [hunk 3]
```diff
@@ -129,7 +130,7 @@ def get_metadata(self) -> Dict[str, Any]:
 
         If no metadata is stored, an empty dict is returned.
         """
-        metadata_path = os.path.join(self.path, _METADATA_FILE_NAME)
+        metadata_path = Path(self.path, _METADATA_FILE_NAME).as_posix()
```
[the remark this answers is not in the document]
Comment (justinvyu) [reply python/ray/train/_checkpoint.py:133]: I was thinking about it but I prefer just seeing this spelled out. I think the longer term goal is to replace everything with `Path` objects and never pass strings around.

File: python/ray/tune/logger/json.py (hunk 2 above)
Comment (justinvyu) [reply python/ray/tune/logger/json.py:44]: I just used `Path.open` instead of `open(str)` below.

**Comment by n30111** [comment]
> https://github.com/ray-project/ray/blob/master/python/ray/train/tests/test_windows.py

This test covers the scenario https://github.com/ray-project/ray/blob/master/python/ray/train/tests/test_trainer_restore.py#L189, is this run on windows system also?

**Comment by n30111** [comment]
> @n3011 Is it possible to add a small test to confirm that this PR fixes the failure scenario you're running into? Thanks!
> 
> https://github.com/ray-project/ray/blob/master/python/ray/train/tests/test_windows.py

This test covers the scenario https://github.com/ray-project/ray/blob/master/python/ray/train/tests/test_trainer_restore.py#L189, is this run windows system also?

**Comment by justinvyu** [comment]
Thanks for this contribution @n30111! 🚀  Let me know if you get a chance to try it out on nightly.

**Pull request merged**